google-cloud-tools-plugin/src/com/google/gct/idea/appengine/cloud/AppEngineCloudType.java

```diff
@@ -196,6 +201,59 @@ public String getPresentableName() {
       public void setName(String name) {
         this.name = name;
       }
+
+      @NotNull
+      @Override
+      public DeploymentSourceType<?> getType() {
+        return DeploymentSourceType.EP_NAME.findExtension(
+            UserSpecifiedPathDeploymentSourceType.class);
+      }
+    }
+  }
+
+  public static class UserSpecifiedPathDeploymentSourceType extends
+      DeploymentSourceType<ModuleDeploymentSource> {
+    private static final String NAME_ATTRIBUTE = "name";
+    private static final String SOURCE_TYPE_ID = "filesystem-war-jar-module";
+
+    public UserSpecifiedPathDeploymentSourceType() { super(SOURCE_TYPE_ID); }
+
+    @NotNull
+    @Override
+    /**
+     * Restore presentable name (e.g., to be "Filesystem JAR or WAR file - <file path>") of
+     * UserSpecifiedPathDeploymentSource.
+     */
+    public ModuleDeploymentSource load(@NotNull Element tag, @NotNull Project project) {
+      AppEngineDeploymentConfigurator.UserSpecifiedPathDeploymentSource userSpecifiedSource =
+          new AppEngineDeploymentConfigurator.UserSpecifiedPathDeploymentSource(
+              ModulePointerManager.getInstance(project).create(
+                  tag.getAttributeValue(NAME_ATTRIBUTE)));
+
+      Element settings = tag.getChild(DeployToServerRunConfiguration.SETTINGS_ELEMENT);
+      if (settings != null) {
+        String filePath = settings.getAttributeValue(
+            AppEngineDeploymentConfiguration.USER_SPECIFIED_ARTIFACT_PATH_ATTRIBUTE);
+
+        if (!StringUtil.isEmpty(filePath)) {
+          userSpecifiedSource.setName(
+              GctBundle.message(
+                  "appengine.flex.user.specified.deploymentsource.name.with.filename",
+                  new File(filePath).getName()));
+
+          return userSpecifiedSource;
+        }
+      }
+
+      userSpecifiedSource.setName(
+          GctBundle.message("appengine.flex.user.specified.deploymentsource.name"));
+
+      return userSpecifiedSource;
```

Comment:
The default name ('appengine.flex.user.specified.deploymentsource.name') should also be set when 'settings == null', so if we put it inside an else branch, the outer if statement should also have the same duplicate else branch. What do you think?